File: eval-dotnet/tests/Eval.Tests/BehaviourTests.cs
using Eval;

namespace Eval.Tests;

/// <summary>Reading runs against each other at a fixed address.</summary>
public class BehaviourTests
{
    private static RunRecord Run(string id, string address, string arrangement, params string[] proposed) =>
        new(RunRecord.FormV1, id, "tool", "subject", "act/a", DateTimeOffset.UnixEpoch,
            "model", "host", 0, proposed, [], "reply", [],
            Pinned.Of(("q", address)), Pinned.Of(("model", arrangement)));

    private static RunRecord Bare(string id) =>
        new(RunRecord.FormV1, id, "tool", "subject", "act/a", DateTimeOffset.UnixEpoch,
            "model", "host", 0, [], [], "reply", []);

    [Fact]
    public void Runs_without_an_address_are_not_compared()
    {
        Assert.Empty(Behaviour.Read([Bare("01")]));
    }

    [Fact]
    public void Runs_at_one_address_are_read_together()
    {
        var readings = Behaviour.Read([Run("01", "q1", "m", "det/a"), Run("02", "q1", "m", "det/a")]);
        Assert.Single(readings);
        Assert.Equal(2, readings[0].Runs);
    }

    [Fact]
    public void Runs_at_different_addresses_are_not_comparable()
    {
        Assert.Equal(2, Behaviour.Read([Run("01", "q1", "m"), Run("02", "q2", "m")]).Count);
    }

    [Fact]
    public void A_repeated_answer_agrees_with_itself()
    {
        var readings = Behaviour.Read([Run("01", "q1", "m", "det/a"), Run("02", "q1", "m", "det/a")]);
        Assert.Equal(1.0, readings[0].Agreement);
    }

    [Fact]
    public void A_changed_answer_at_one_address_disagrees()
    {
        var readings = Behaviour.Read([Run("01", "q1", "m", "det/a"), Run("02", "q1", "m", "det/b")]);
        Assert.Equal(0.0, readings[0].Agreement);
    }

    /// <summary>Two runs that both proposed nothing agree. Silence is a real answer.</summary>
    [Fact]
    public void Two_silences_agree()
    {
        var readings = Behaviour.Read([Run("01", "q1", "m"), Run("02", "q1", "m")]);
        Assert.Equal(1.0, readings[0].Agreement);
    }

    /// <summary>One run at an address is not a low score; it is the absence of one.</summary>
    [Fact]
    public void A_single_run_yields_no_agreement_rather_than_zero()
    {
        var readings = Behaviour.Read([Run("01", "q1", "m", "det/a")]);
        Assert.Null(readings[0].Agreement);
        Assert.Null(readings[0].Drift);
    }

    /// <summary>The earthquake reading: same address, changed worker, moved answer.</summary>
    [Fact]
    public void A_changed_arrangement_at_a_fixed_address_shows_as_drift()
    {
        var readings = Behaviour.Read([
            Run("01", "q1", "old", "det/a"), Run("02", "q1", "old", "det/a"),
            Run("03", "q1", "new", "det/b"), Run("04", "q1", "new", "det/b"),
        ]);

        Assert.Equal(2, readings[0].Arrangements);
        Assert.Equal(1.0, readings[0].Agreement);
        Assert.Equal(0.0, readings[0].AcrossArrangements);
        Assert.Equal(1.0, readings[0].Drift);
    }

    [Fact]
    public void A_stable_worker_change_shows_no_drift()
    {
        var readings = Behaviour.Read([
            Run("01", "q1", "old", "det/a"), Run("02", "q1", "old", "det/a"),
            Run("03", "q1", "new", "det/a"), Run("04", "q1", "new", "det/a"),
        ]);
        Assert.Equal(0.0, readings[0].Drift);
    }

    /// <summary>Declared-distinct ground must yield distinct behaviour.</summary>
    [Fact]
    public void Two_addresses_answered_identically_are_reported_as_collapsed()
    {
        var collapse = Behaviour.Collapsed([Run("01", "q1", "m", "det/a"), Run("02", "q2", "m", "det/a")]);
        Assert.Single(collapse);
    }

    [Fact]
    public void Addresses_answered_differently_have_not_collapsed()
    {
        Assert.Empty(Behaviour.Collapsed([Run("01", "q1", "m", "det/a"), Run("02", "q2", "m", "det/b")]));
    }

    /// <summary>An address answered two ways has not collapsed onto another.</summary>
    [Fact]
    public void An_unstable_address_is_not_reported_as_collapsed()
    {
        var collapse = Behaviour.Collapsed([
            Run("01", "q1", "m", "det/a"), Run("02", "q1", "m", "det/b"), Run("03", "q2", "m", "det/a"),
        ]);
        Assert.Empty(collapse);
    }

    /// <summary>The two runtimes must read the same runs the same way.</summary>
    [Fact]
    public void An_address_pins_the_same_digest_as_the_rust_half_would()
    {
        Assert.Equal(Pinned.Of(("q", "q1")).Digest, Pinned.Of(("q", "q1")).Digest);
        Assert.NotEqual(Pinned.Of(("q", "q1")).Digest, Pinned.Of(("q", "q2")).Digest);
    }
}

File: eval-core/src/behaviour_tests.rs
use super::*;

use crate::pinned::Pinned;

fn run(id: &str, address: &str, arrangement: &str, proposed: &[&str]) -> RunRecord {
    let mut record = RunRecord::new(id, "tool", "act/a", "subject")
        .at(Pinned::of([("q", address)]))
        .arranged_as(Pinned::of([("model", arrangement)]));
    record.proposed = proposed.iter().map(|s| (*s).to_string()).collect();
    record
}

#[test]
fn runs_without_an_address_are_not_compared() {
    let bare = RunRecord::new("01", "tool", "act/a", "subject");
    assert!(read(&[bare]).is_empty());
}

#[test]
fn runs_at_one_address_are_read_together() {
    let readings = read(&[run("01", "q1", "m", &["det/a"]), run("02", "q1", "m", &["det/a"])]);
    assert_eq!(readings.len(), 1);
    assert_eq!(readings[0].runs, 2);
}

#[test]
fn runs_at_different_addresses_are_not_comparable() {
    assert_eq!(read(&[run("01", "q1", "m", &[]), run("02", "q2", "m", &[])]).len(), 2);
}

#[test]
fn a_repeated_answer_agrees_with_itself() {
    let readings = read(&[run("01", "q1", "m", &["det/a"]), run("02", "q1", "m", &["det/a"])]);
    assert_eq!(readings[0].agreement, Some(1.0));
}

#[test]
fn a_changed_answer_at_one_address_disagrees() {
    let readings = read(&[run("01", "q1", "m", &["det/a"]), run("02", "q1", "m", &["det/b"])]);
    assert_eq!(readings[0].agreement, Some(0.0));
}

/// Two runs that both proposed nothing agree. Silence is a real answer.
#[test]
fn two_silences_agree() {
    let readings = read(&[run("01", "q1", "m", &[]), run("02", "q1", "m", &[])]);
    assert_eq!(readings[0].agreement, Some(1.0));
}

/// One run at an address is not a low score; it is the absence of one.
#[test]
fn a_single_run_yields_no_agreement_rather_than_zero() {
    let readings = read(&[run("01", "q1", "m", &["det/a"])]);
    assert_eq!(readings[0].agreement, None);
    assert_eq!(readings[0].drift(), None);
}

/// The earthquake reading: same address, changed worker, moved answer.
#[test]
fn a_changed_arrangement_at_a_fixed_address_shows_as_drift() {
    let readings = read(&[
        run("01", "q1", "old", &["det/a"]),
        run("02", "q1", "old", &["det/a"]),
        run("03", "q1", "new", &["det/b"]),
        run("04", "q1", "new", &["det/b"]),
    ]);

    assert_eq!(readings[0].arrangements, 2);
    assert_eq!(readings[0].agreement, Some(1.0), "each worker agrees with itself");
    assert_eq!(readings[0].across_arrangements, Some(0.0), "they do not agree with each other");
    assert_eq!(readings[0].drift(), Some(1.0));
}

#[test]
fn a_stable_worker_change_shows_no_drift() {
    let readings = read(&[
        run("01", "q1", "old", &["det/a"]),
        run("02", "q1", "old", &["det/a"]),
        run("03", "q1", "new", &["det/a"]),
        run("04", "q1", "new", &["det/a"]),
    ]);
    assert_eq!(readings[0].drift(), Some(0.0));
}

/// Declared-distinct ground must yield distinct behaviour.
#[test]
fn two_addresses_answered_identically_are_reported_as_collapsed() {
    let collapse = collapsed(&[run("01", "q1", "m", &["det/a"]), run("02", "q2", "m", &["det/a"])]);
    assert_eq!(collapse.len(), 1, "the caller declared two questions and got one answer");
}

#[test]
fn addresses_answered_differently_have_not_collapsed() {
    assert!(collapsed(&[run("01", "q1", "m", &["det/a"]), run("02", "q2", "m", &["det/b"])]).is_empty());
}

/// An address that answered two different ways has not collapsed onto another.
#[test]
fn an_unstable_address_is_not_reported_as_collapsed() {
    let collapse = collapsed(&[
        run("01", "q1", "m", &["det/a"]),
        run("02", "q1", "m", &["det/b"]),
        run("03", "q2", "m", &["det/a"]),
    ]);
    assert!(collapse.is_empty());
}
